### PR TITLE
Week 5: Calendar Project

### PR DESCRIPTION
**Pull request by danieldortega77**
Complete the query() method in FindMeetingQuery.java, and add functionality for optional attendees.

Write test cases that check the functionality of optional attendees.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,9 +15,95 @@
...
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> output = Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
```
Comment (dborkan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I'd recommendation changing variable names like `output`, `mandatoryOutput` etc to be a little more descriptive.  Maybe something like `availableTimeRanges`?  Or even `availableTimes`?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,95 @@
...
+    // For each time slot of that attendee
+    for (Event event : events) {
+      if (event.getAttendees().contains(attendee)) {
+        TimeRange existingTR = event.getWhen();
+        List<TimeRange> tempRanges = new ArrayList<TimeRange>();
+
+        // For each currently "open" time slot
+        for (TimeRange possibleTR : output) {
+          if (possibleTR.overlaps(existingTR)) {
+            // Case 1: E |------|
```
Comment (dborkan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: Nice diagram :)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,95 @@
...
+
+    List<TimeRange> optionalOutput = cleanUpOutput(request, output);
+
+    if (optionalOutput.size() == 0 && request.getAttendees().size() != 0) {
+      return mandatoryOutput;
+    }
+    return optionalOutput;
+  }
+
+  List<TimeRange> updateOutput(Collection<Event> events, String attendee, List<TimeRange> output) {
```
Comment (dborkan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Also maybe change this name to something more descriptive.  I'm thinking maybe `findAvailableTimeRanges` or something similar?

**Review by dborkan — APPROVED** [review]
👍 Some nits on variable and function names, but otherwise the algorithm and tests look good!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (danieldortega77) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yes, that makes sense.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (danieldortega77) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: Thanks!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (danieldortega77) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Will do!

**Review by dslucas — APPROVED** [review]

**Pull request merged**